src/transformStateWithClones.js

```diff
@@ -1,13 +1,40 @@
 'use strict';
-
-/**
+/*
  * @param {Object} state
  * @param {Object[]} actions
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const updateState = { ...state };
+  const updateActions = [];
+
+  for (const key of actions) {
+    switch (key.type) {
+      case 'addProperties':
+        Object.assign(updateState, key.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const item of key.keysToRemove) {
+          delete updateState[item];
+        }
+        break;
+
+      case 'clear':
+        for (const a in updateState) {
+          delete updateState[a];
+        }
+        break;
+
+      default:
+        break;
+    }
+    updateActions.push({ ...updateState });
```

Comment:
[DONT REPEAT YOURSELF]: Pushing the copy of the state to the `updateActions` array is done correctly outside the switch block. This is good because it avoids repetition and ensures the action is performed only once per iteration.